### PR TITLE
Startup update

### PR DESCRIPTION
**Pull request by ssakone**
### Login update step
- [x] Startup Page
- [x] Wallet Login
- [x] New Wallet registration
- [x] Smart Seed verification
- [x] Wallet Restoration
- [x] Langue selection widget
- [x] Loading ui

image
image
image

### CODE REVIEW

**Comment by Milerius** [comment]
> ### Login update step
> * [x]  Startup Page
> * [x]  Wallet Login
> * [ ]  New Wallet registration
> * [ ]  Wallet Restoration
> * [x]  Loading ui
> 
> image
> image
> image

Got a first warning: `[06:41:26] [warning] [main.prerequisites.hpp:126] [5020275]: qrc:/atomic_defi_design/qml/Components/SetupPage.qml:15:5: Unable to assign [undefined] to QString
`

File: atomic_defi_design/qml/Components/DexAppButton.qml [hunk 1]
```diff
@@ -1,19 +1,49 @@
...
-        id: _label
-        anchors.centerIn: parent
-        font: _font.button
-        text: parent.text
+    Row {
+        id: _contentRow
+        anchors {
+            //centerIn: parent.alignment == Qt.AlignCenter? parent : undefined 
+            right: parent.horizontalAlignment == Qt.AlignRight? parent.right : undefined
+            left: parent.horizontalAlignment == Qt.AlignLeft? parent.left : undefined
```
Comment (Milerius) [inline atomic_defi_design/qml/Components/DexAppButton.qml:30]: Coding style

File: atomic_defi_design/qml/Components/DexAppButton.qml [hunk 2]
```diff
@@ -1,19 +1,49 @@
...
+            horizontalCenter: parent.horizontalAlignment == Qt.AlignHCenter ? parent.horizontalCenter : undefined
+            verticalCenter: parent.verticalAlignment == Qt.AlignVCenter ? parent.verticalCenter :  undefined
+        }
+        spacing: parent.spacing
+        Qaterial.ColorIcon {
+            id: _icon 
+            anchors.verticalCenter: parent.verticalCenter
+            iconSize: _label.font.pixelSize+2
```
Comment (Milerius) [inline atomic_defi_design/qml/Components/DexAppButton.qml:38]: coding style

File: atomic_defi_design/qml/Components/DexAppTextField.qml [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	property int leftWidth: -1
+	readonly property int max_length: 40 
+
+	Rectangle {
+		id: _background
+	    anchors.fill: parent
+	    radius: 4
+	    color: theme.surfaceColor
+	    border.color: theme.accentColor
+	    border.width: input_field.focus? 1 : 0
```
Comment (Milerius) [inline atomic_defi_design/qml/Components/DexAppTextField.qml:24]: coding style

File: atomic_defi_design/qml/Components/DexAppTextField.qml [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+	}
+
+	RowLayout {
+		anchors.fill: parent
+		anchors.leftMargin: 5
+		anchors.rightMargin: 5
+		spacing: 2
+		Item {
+			visible: leftText!==""
+			Layout.preferredWidth: leftWidth!==-1? leftWidth : _title_label.implicitWidth+2
```
Comment (Milerius) [inline atomic_defi_design/qml/Components/DexAppTextField.qml:34]: coding style

File: atomic_defi_design/qml/Components/DexAppTextField.qml [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+			        background: Item{}
+			        font.weight: Font.Medium
+			        font.family: 'Lato'
+			        font.pixelSize: 13
+			        anchors.fill: parent
+			    }
+			}
+		}
+		Item {
+			visible: rightText!==""
```
Comment (Milerius) [inline atomic_defi_design/qml/Components/DexAppTextField.qml:77]: coding style

File: atomic_defi_design/qml/Screens/FirstLaunch.qml [hunk 6]
```diff
@@ -61,60 +68,98 @@ SetupPage {
...
                             anchors.fill: parent
-                            hoverEnabled: true
-                            onClicked: {
-                                selected_wallet_name = model.modelData
-                                onClickedWallet()
+
+                            // Click area
+                            Rectangle {
+                                height: parent.height
+                                width: mouse_area.containsMouse? parent.width : 0
```
Comment (Milerius) [inline atomic_defi_design/qml/Screens/FirstLaunch.qml:118]: coding style

File: atomic_defi_design/qml/Screens/InitialLoading.qml [hunk 7]
```diff
@@ -21,24 +21,28 @@ SetupPage {
...
+    backgroundColor: 'transparent'
+    borderColor: 'transparent'
     content: ColumnLayout {
+
+        DefaultBusyIndicator {
+            Layout.preferredHeight: 100
+            Layout.preferredWidth: 100 
+            Layout.alignment: Qt.AlignHCenter
+            Layout.leftMargin: -15
+            Layout.rightMargin: Layout.leftMargin*0.75
```
Comment (Milerius) [inline atomic_defi_design/qml/Screens/InitialLoading.qml:33]: coding style

File: atomic_defi_design/qml/Screens/Login.qml [hunk 8]
```diff
@@ -53,40 +55,180 @@ SetupPage {
...
                 id: submit_button
-                Layout.fillWidth: true
-                implicitHeight: _back.implicitHeight
                 text: qsTr("Login")
-                onClicked: trySubmit()
                 enabled: input_password.isValid()
+                onClicked: trySubmit()
+                radius: 20
+                backgroundColor: theme.accentColor
+                Layout.preferredWidth: _nextRow.implicitWidth+40
```
Comment (Milerius) [inline atomic_defi_design/qml/Screens/Login.qml:213]: coding style

**Review by Milerius — CHANGES_REQUESTED** [review]

File: atomic_defi_design/qml/Screens/InitialLoading.qml (hunk 7 above)
Comment (ssakone) [reply atomic_defi_design/qml/Screens/InitialLoading.qml:33]: old code

**Comment by smk762** [comment]
Login with existing user working ok.
image

Not sure why the user left section / icon is faded compared to the password icon?
Or what the purpose of this bit between the field is - 
image

**Comment by ssakone** [comment]
> Login with existing user working ok.
> image
> 
> Not sure why the user left section / icon is faded compared to the password icon?
> Or what the purpose of this bit between the field is -
> image

User field is not is not editable and i belive i need to change icon, because its not user but wallet

**Comment by tonymorony** [comment]
Great progress, current feedback:

1) Don't allow multiline on seed phrase input, that's potentially very dangerous (user will generate different wallet from his seed if even single `/n` will be added):

Screenshot 2021-07-04 at 03 26 56

2)  Show only 4 words as an option here, otherwise, it's too much time to find a needed word in a huge list and it's faster just type:

image

3) There is no login button at the moment and no indication that the user input not correct password

image

**Review by tonymorony — DISMISSED** [review]
left feedback from the testing session above, please pay special attention to seed validation it's a critical component of the app

**Review by Milerius — CHANGES_REQUESTED** [review]
After the first try - 2nd word of seed was asked twice during the seed validation.

**Comment by Milerius** [comment]
Few remarks

- Custom seed should allow any type of seed
- When custom seed is not ticked the valid seed are either 12 or 24 words seed
- If a default wallet is already selected, clicking on return goes back to this specific page


https://user-images.githubusercontent.com/21139416/124472552-8e7c1580-dd9e-11eb-9751-c52cb20b1460.mov

**Comment by smk762** [comment]
> 1. I think it might be worth adding wallet name somewhere on this page
> 
> image

Something like this would be nice
image

**Comment by smk762** [comment]
Re: https://github.com/KomodoPlatform/atomicDEX-Desktop/actions/runs/1001035197

- Confirm newline in wallet restore textarea issue resolved (both custom and non-custom) :heavy_check_mark: 

- non-custom seed validation incomplete. :stop_sign: 
  - should not be allowing leading space before first word (ideally not for custom seed either)
  - should not be allowing trailing space after last word (ideally not for custom seed either)
  - should not be allowing capitals as input
  - should not be allowing special chars as input
  - should not be allowing non 12/24 word count (if trailing space)
  - should not be allowing words are not in BIP39 wordlist
  
image

  - Word selection for new seed 4 word choice confirmation :heavy_check_mark: 
    - **BUT** user can just spam selection until they get it right, i.e. not forced to write down.
    - Suggest failed attempt changes number of word to be confirmed, and 3 failed attempts aborts wallet creation.

https://user-images.githubusercontent.com/35845239/124552797-91423d80-de66-11eb-82ac-6024b69aa476.mp4
  
> ...no indication that the user input not correct password
- Confirm failed password effect (shake and color change) triggered on "connect" button press if incorrect. :heavy_check_mark: 

> There is a bug if click on the same translation flag twice:
- Confirm resolved :heavy_check_mark: 

> If a default wallet is already selected, clicking on return goes back to this specific page 
- Unable to replicate, assume resolved.

**Comment by ssakone** [comment]
- Enhance Seed Verification & improve user interaction
- Add Bip39Wordlist Checking
- Change login screen display to new splash login
- improve login page

File: atomic_defi_design/qml/Screens/RecoverSeed.qml [hunk 9]
```diff
@@ -66,98 +86,309 @@ SetupPage {
...
             }
 
-            hidable: true
-            hiding: false
-            hide_button.use_default_behaviour: false
-            hide_button_area.onClicked: {
-                visible = false
-                // input_seed_hidden.field.focus = true // This puts the cursor to left, not good
+            DexBip39WordListChecker {
+                id: big39Checker
```
Comment (tonymorony) [inline atomic_defi_design/qml/Screens/RecoverSeed.qml:178]: it's a bip, not big :)

**Comment by smk762** [comment]
Password where visible is slightly too large for input area - suggest slight font size reduction
image

Appears that incorrect password no longer alerts user when input
```
[13:08:46] [info] [app.cpp:644] [540000]: Application active
[13:08:50] [info] [qt.wallet.manager.cpp:271] [540000]: qt_wallet_manager::login
[13:08:50] [info] [qt.wallet.manager.cpp:183] [540000]: Loading wallet configuration: cswm
[13:08:50] [info] [qt.wallet.manager.cpp:197] [540000]: wallet_cfg: {
    "name": "cswm"
}
[13:08:50] [debug] [security.utilities.cpp:52] [540000]: derive_password l52 f[security.utilities.cpp]
[13:08:50] [info] [security.utilities.cpp:65] [540000]: Key generated successfully
[13:08:50] [debug] [security.utilities.cpp:100] [540000]: decrypt l100 f[security.utilities.cpp]
[13:08:50] [warning] [qt.wallet.manager.cpp:313] [540000]: Wrong password, or the seed file is corrupted.
[13:08:50] [debug] [main.prerequisites.hpp:120] [540000]: Failed: Login
[13:08:50] [warning] [main.prerequisites.hpp:126] [540000]: qrc:/atomic_defi_design/qml/Screens/FirstLaunch.qml:32: Error: Invalid write to global property "text_error"
```

When entering password, but then cancelling and returning to select a different wallet, the previously entered password is still in the input. Hidden/Shown state is also persisting. On selection of a wallet, this input should be reset to empty and hidden.

https://user-images.githubusercontent.com/35845239/124703975-e09b7300-df25-11eb-9ed9-fd5d632a4d8f.mp4

**Comment by smk762** [comment]
image

`Incorrect seed try again` is ambiguous. Suggest `BIP39 seed validation failed, try again or select 'Allow custom seed'`

**Comment by Milerius** [comment]
Please remove the word checker list and used the function that is present for mnemonic validation: 

`Q_INVOKABLE static bool        mnemonic_validate(const QString& entropy);`

`./qml/Screens/RecoverSeed.qml:157:                             (allow_custom_seed.checked || API.app.wallet_mgr.mnemonic_validate(input_seed.field.text))`

File: atomic_defi_design/qml/Components/DexBip39WordListChecker.qml [hunk 10]
```diff
@@ -0,0 +1,2061 @@
+import QtQuick 2.15
```
Comment (tonymorony) [inline atomic_defi_design/qml/Components/DexBip39WordListChecker.qml:1]: there is mnemonic_validate function for that https://github.com/KomodoPlatform/atomicDEX-Desktop/blob/dev/atomic_defi_design/qml/Screens/RecoverSeed.qml#L157

**Comment by smk762** [comment]
Slight button overflow when error message appears on wallet recover- 
image

Only happens on first try, subsequent attempts the button is within panel.

**Comment by smk762** [comment]
`Cotinue` should be `Continue`
image

**Comment by smk762** [comment]
"View Seed" for newly created wallet is incomplete (created thru "new wallet" button).

image

Checked older wallet, full seed was on display. Restored wallet (with custom seed) also displayed ok.

**Comment by Milerius** [comment]
> "View Seed" for newly created wallet is incomplete (created thru "new wallet" button).
> 
> image
> 
> Checked older wallet, full seed was on display. Restored wallet (with custom seed) also displayed ok.
> 
> **Very lucky I caught this**
> I had not thought to check that until today. Anything related to seed / passphrase / private key needs to be handled with care and checked more often than Santa Claus' list.
> 
> If merged and found by a user unable to recover account, or running an account with a single word seed which was hacked, there would be some serious egg on our face.
> 
> In this case, I checked using `life` as a custom seed - it gave same address, which makes it a massively insecure wallet.
> 
> Perhaps we should implement some standards for custom seeds so that there is no chance of a seed less than a certain level of entropy.

I do agree on wallet creation for additional security check, like internal validate from my side (i was expecting that `mnemonic_validate` will be called before final validation by front end, because in this case we can show an error in the front-end, that's why this function was exposed in a nice way in order to be used in front end even before the wallet creation, not only on restore - pretty sure it's was the case before.

File: atomic_defi_design/qml/Components/DexAppTextArea.qml [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+import QtQuick.Controls 2.15
+import Qaterial 1.0 as Qaterial
+import QtQuick.Layouts  1.5
+
+Item {
+	id: control
+	width:  200
+    height: 35
+    signal accepted()
+    property alias value: input_field.text
```
Comment (Milerius) [inline atomic_defi_design/qml/Components/DexAppTextArea.qml:11]: coding style / alignment

File: atomic_defi_design/qml/Components/DexAppTextField.qml [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+import Qaterial 1.0 as Qaterial
+import QtQuick.Layouts  1.5
+
+import App 1.0
+
+Item {
+	id: control
+	width:  200
+    height: 35
+    property alias value: input_field.text
```
Comment (Milerius) [inline atomic_defi_design/qml/Components/DexAppTextField.qml:12]: coding style / alignment

File: atomic_defi_design/qml/Components/DexAppTextField.qml [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+			} else {
+				_background.x = -3
+			}
+		}
+	}
+	function reset() {
+		input_field.text = ""
+	}
+	Rectangle {
+		id: _background
```
Comment (Milerius) [inline atomic_defi_design/qml/Components/DexAppTextField.qml:50]: coding style / alignment

File: atomic_defi_design/qml/Screens/RecoverSeed.qml [hunk 14]
```diff
@@ -10,6 +12,7 @@ SetupPage {
     // Override
     property var onClickedBack: () => {}
     property var postConfirmSuccess: () => {}
```
Comment (Milerius) [inline atomic_defi_design/qml/Screens/RecoverSeed.qml:14]: postConfirmSuccess should clean the seed and the password from memory.

**Comment by smk762** [comment]
Sometimes when logging out, It goes back to this screen

image

no wallet name, so not sure what I'm logging in to.

This only happens sometimes, have been unable to catch exact steps to replicate it every time.

**Comment by Milerius** [comment]
> Sometimes when logging out, It goes back to this screen
> 
> image
> 
> no wallet name, so not sure what I'm logging in to.
> 
> This only happens sometimes, have been unable to catch exact steps to replicate it every time. Should return to the wallet selection page every time.

Yeah it's should always go back to the wallet selection not this page, good catch

**Comment by ca333** [comment]
@ssakone please do validate and test your code **BEFORE** commit/push!

**Comment by ssakone** [comment]
> @ssakone please do validate and test your code **BEFORE** commit/push!

I'm sorry for the inconvenience and will be even more careful.

**Comment by Milerius** [comment]
> Slight button overflow when error message appears on wallet recover-
> image
> 
> Only happens on first try, subsequent attempts the button is within panel.

Let's check this overflow and the logout problem so we can maybe merge it afterwards @ssakone if you have any doubt or question please ask on the keybase channel

**Comment by ssakone** [comment]
@tonymorony 
image
about when user have lot of wallet like me, is there anything to delete some wallet directly or will you have to connect to each one to delete?

**Comment by Milerius** [comment]
> @tonymorony
> image
> about when user have lot of wallet like me, is there anything to delete some wallet directly or will you have to connect to each one to delete?

could be done into a separate pr

**Comment by smk762** [comment]
overflow button on nonBIP39 seed fixed :heavy_check_mark: 
new wallet full seed phrase issue fixed :heavy_check_mark: 
`continue` spelling fixed :heavy_check_mark: 
Password not clearing if cancelling to select different wallet :crab: 

https://user-images.githubusercontent.com/35845239/125156999-01b8c980-e19b-11eb-9b43-077d839d6dc4.mp4

**Comment by smk762** [comment]
Password clearing when cancelling to select different wallet :heavy_check_mark:

File: atomic_defi_design/qml/Screens/NewUser.qml [hunk 15]
```diff
@@ -277,37 +436,284 @@ SetupPage {
...
+                Layout.preferredHeight: 50
+                background.border.width: 1
+                background.radius: 25
+                background.border.color: field.focus ? theme.accentColor : Style.colorBorder
+                field.echoMode: TextField.Password
+                field.font: field.echoMode === TextField.Password ? field.text === "" ? theme.textType.body1 : theme.textType.head5 : theme.textType.head6
+                field.horizontalAlignment: Qt.AlignLeft
+                field.leftPadding: 75
+                field.rightPadding: 60
+                field.placeholderText: qsTr("Cofirm password")
```
Comment (Milerius) [inline atomic_defi_design/qml/Screens/NewUser.qml:636]: Confirm password

**Review by smk762 — DISMISSED** [review]

**Review by Milerius — APPROVED** [review]

**Comment by Milerius** [comment]
Final review @smk762 @tonymorony plz

**Review by smk762 — APPROVED** [review]

**Comment by tonymorony** [comment]
is it intended that these icons are blured and unclickable?

Screenshot 2021-07-12 at 16 54 00

**Comment by Milerius** [comment]
> is it intended that these icons are blured and unclickable?
> 
> Screenshot 2021-07-12 at 16 54 00

Yes, it's the effect

**Review by tonymorony — APPROVED** [review]

**Pull request merged**